### PR TITLE
Iteration Slides

### LINKED ISSUE #10: Slides: Iteration
Sources:

- [Intermediate class](https://github.com/uc-python/intermediate-python-datasci/blob/master/notebooks/03-Iteration.ipynb)
- [BANA 6043](https://github.com/bradleyboehmke/uc-bana-6043/blob/main/book/06-module/lesson-6b.ipynb)

### PR DESCRIPTION
**Pull request by augustopher**
Closes #10 

Adds the Iteration materials from the [intermediate class](https://github.com/uc-python/intermediate-python-datasci/blob/master/notebooks/03-Iteration.ipynb), with some edits:

- Removes pandas content and examples
- Replaces some pandas examples with plain Python examples

Some questions to consider for potential changes:

- The intermediate class doesn't cover dictionaries or dictcomps at all, and only passing references to both are made in the 6043 class. Do we want to supplement this material with dictionary iterations (likely following the `enumerate` section) and dictcomps (a quick example or two after listcomps)?
- I noticed some HTML formatting that seemed to be tied to [`rise.css`](https://github.com/uc-python/intermediate-python-datasci/blob/master/notebooks/rise.css) in the intermediate class - do we still want that formatting, and how is it integrated into the slide generation?
- Should we also have a quick CI step to check that the slides build successfully? I doubt it will be a problem, but could be a nice basic and passive way check. Still doesn't mean the slides look right, but at least there's no egregious errors in building

### CODE REVIEW

**Comment by bradleyboehmke** [comment]
If you wanted to you could add a slide that illustrates dictionary comprehensions. I would think once they have list comprehensions down that will be an easy extension to grasp.

As for the rise.css, I would remove that content.  We haven't used rise for a while now and we mainly just produce the HTML slides and use those.

**Comment by eswan18** [comment]
Agree with Brad on both of those statements.

> Should we also have a quick CI step to check that the slides build successfully? I doubt it will be a problem, but could be a nice basic and passive way check. Still doesn't mean the slides look right, but at least there's no egregious errors in building.

I can't ever recall, even once, getting an error when using `nbconvert` to build slides. That's not to say it's impossible, but I guess I always assumed that running it as a CI step was of dubious value. Not even knowing what errors look like, I'm not sure how we'd check for "success" except for confirming that the shell exit code was 0.

**Comment by eswan18** [comment]
And a followup to your question about formatting: I actually think that if we have time before the first class, putting together a little CSS file could be nice, but I don't want to prioritize it over content. Here are the things that come to mind as styling fixes:
- images have really inconsistent sizes and sometimes run off the slide
- `h4` elements (headings with four `#`s) are smaller than body text, which looks silly

File: notebooks/05-Iteration.ipynb [hunk 1]
```diff
@@ -0,0 +1,1565 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
```
Comment (eswan18) [inline notebooks/05-Iteration.ipynb:499]: I believe this cell should begin a new slide. It seems out-of-place with the content before it:
image

**Review by eswan18 — APPROVED** [review]
These look great. I left one suggested change but it's your call.

Very glad you added the section on dictionary comprehensions. Since these students will be mostly developers, I think the majority of this content will be a breeze for them. The only Python-specific ideas are list and dict comprehensions, so it's a good idea to try to move quickly through the beginning and reserve some time for that part.

**Comment by augustopher** [comment]
> I can't ever recall, even once, getting an error when using `nbconvert` to build slides. That's not to say it's impossible, but I guess I always assumed that running it as a CI step was of dubious value. Not even knowing what errors look like, I'm not sure how we'd check for "success" except for confirming that the shell exit code was 0.

Fair enough. I hadn't thought through the actually errors and such, more just adding another safety check. I'm not too worried about it.

> * `h4` elements (headings with four `#`s) are smaller than body text, which looks silly

I was a little confused on the use of `h4` vs `h3` for slide titles. While going through the Functions slides, I thought about just making them all `h3`, given the lack of contrast for `h4` - thoughts?

**Comment by eswan18** [comment]
> I was a little confused on the use of h4 vs h3 for slide titles. While going through the Functions slides, I thought about just making them all h3, given the lack of contrast for h4 - thoughts?

Yep, funny you mention that – I've started doing the same thing in the 01 notebook. I don't think we should go out of our way to fix every single instance of an h4, but it's probably best to switch them to h3s where possible. The use of different heading-levels was never consistent anyway, and it's hard to keep it consistent in these long decks, so defaulting to h3s seems fine.

File: notebooks/05-Iteration.ipynb (hunk 1 above)
Comment (augustopher) [reply notebooks/05-Iteration.ipynb:499]: Agreed. I'll split those up.

**Comment by bradleyboehmke** [comment]
> > I was a little confused on the use of h4 vs h3 for slide titles. While going through the Functions slides, I thought about just making them all h3, given the lack of contrast for h4 - thoughts?
> 
> Yep, funny you mention that – I've started doing the same thing in the 01 notebook. I don't think we should go out of our way to fix every single instance of an h4, but it's probably best to switch them to h3s where possible. The use of different heading-levels was never consistent anyway, and it's hard to keep it consistent in these long decks, so defaulting to h3s seems fine.

Agree.  I remember there being waaaaaaayyyyyyy to many headers and types of headers throughout so clean up is warranted.

**Pull request merged**